### ISSUE TITLE
Feature Request: Allow Arbitrary Environment Variables on ThanosSpec

### ISSUE THREAD

**Issue by Capitrium** [coreos/prometheus-operator#2358]
#2264 removed the ability to reference specific keys of specific secrets for credentials, i.e.:
```yaml
accessKey:
  key: access_key
  name: thanos-vault-aws
secretKey:
  key: secret_key
  name: thanos-vault-aws
```

While this change maintains compatibility with Thanos, it breaks compatibility with services that automate the creation and renewal of credentials, such as [Vault](https://www.vaultproject.io/).

Rather than requiring users to hard-code their credentials in the Thanos config, users should be able to create custom environment variables for the Thanos pods, so that such secrets can continue to be referenced in the `thanos.objectStorageConfig` secret, i.e.:
```yaml
- args:
  - |
    --objstore.config=
    type: S3
    config:
      access_key: "$(AWS_ACCESS_KEY)"
      secret_key: "$(AWS_SECRET_KEY)"
      ...
  env:
  - name: AWS_ACCESS_KEY
    valueFrom:
      secretKeyRef:
        key: access_key
        name: thanos-vault-aws
  - name: AWS_SECRET_KEY
    valueFrom:
      secretKeyRef:
        key: secret_key
        name: thanos-vault-aws
```

**Comment by brancz** [comment]
I'm not sure I understand. Environment variables are not changed and reloaded at runtime, so how is having the config being in the secret different from the credentials directly? Some tool must be writing the provisioning the credentials into the secret from vault, then that should just as easily be able to template the configuration.

**Comment by Capitrium** [comment]
> then that should *just as easily* be able to template the configuration.

I disagree with you here. The effort required and complexity introduced by a templating engine to the custom `vault-controller` provisioning these secrets seems far greater to me than adding an `env` field to the `ThanosSpec`.

Having said that, it would be great to get all secrets from my `vault-controller` to automatically reload on rotation, and being able to mount entire templated config files would address that pretty well. So if you're strongly against adding an `env` field I guess I'll have to start working on that templating functionality.

**Comment by brancz** [comment]
I'd prefer if we did things that way, plus changing the secret itself doesn't redeploy the Thanos pods, so it wouldn't actually solve the problem, you need some shim either way.

**Comment by Capitrium** [comment]
If I'm mounting a templated credentials file as a volume then I wouldn't need to redeploy the Thanos pods at all, unless Thanos only reads the credentials file when it first launches.

In any case, I'll close this and work on the templating functionality upstream. Hopefully I'll find some time to work on it this month... 😆 😭

**Comment by TheKangaroo** [comment]
With #2445 in the 0.30.0 release you can inject env vars in the thanos sidecar with:
```
spec:
  containers:
  - name: "thanos-sidecar"
    env:
    - name: MYENVVAR
      value: MYVALUE

```

**Comment by brancz** [comment]
Note that that is only meant as a last resort, and when you do that we won’t help debugging your bugs. It says this clearly in the docs. If there is a legit case to implement some configuration please still open an issue

**Comment by TheKangaroo** [comment]
@brancz, I posted my usecase here in the slack channel:
https://kubernetes.slack.com/archives/CFFDS2Z7F/p1557309118156400

*tl;dr: is it possible to pass env vars to the thanos sidecar container?*
Hey guys, we have a setup where we create multiple prometheus with thanos sidecars from the operator. each prometheus/thanos stack gets its own s3 bucket automaticaly via the aws service broker. But this results in a setup where we have a secret with the access key/secret key (created by the service broker) and another secret with the rest of the s3 config. For thanos itself it works fine to pass the access and secret key via env vars to the container and the rest of the config via the file.
Unfortunateley the prometheus crd does no longer supports access and secret key via env vars from a secret since the s3 key is deprecated and objectStorageConfig replaces it. From the crd description this was possible with the s3 config till thanos version v0.2.0
Or am I missing something on how I can pass env vars to the thanos sidecar container?
